include/nbl/builtin/glsl/shapes/rectangle.glsl

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_GLSL_SHAPES_RECTANGLE_INCLUDED_
+#define _NBL_BUILTIN_GLSL_SHAPES_RECTANGLE_INCLUDED_
+
+#include <nbl/builtin/glsl/math/functions.glsl>
+
+struct SphRect {
+    mat3 basis;
+    vec3 r0;
+    vec3 r1;
+};
+
+SphRect nbl_glsl_shapes_getSphericalRectangle(in vec3 start, in vec3 ex, in vec3 ey, in vec3 origin) 
+{
+    const float exl = length(ex);
+    const float eyl = length(ey);
+    const vec3 x = ex / exl;
+    const vec3 y = ey / eyl;
+    vec3 z = normalize(cross(x, y));
+    if (dot(start-origin, z) > 0) {
+        z*=-1;
+    }
+
+    const mat3 basis = mat3(x, y, z);
+    const vec3 r0 = (start-origin) * basis;
```

Comment:
you can flip `z` (last of column of `basis`) and `r0.z` AFTER you compute `r0` via `(start-origin)*basis`

```glsl
const uint signFlipMask = (floatBitsToUint(r0.z)^0x80000000u)&0x80000000u;
basis[3] = uintBitsToFloat(floatBitsToUint(basis[3])^signFlipMask);
r0.z = uintBitsToFloat(floatBitsToUint(r0.z)^signFlipMask);
```

Reply:
actually you don't even need to keep `basis` around after computing `r0`

Reply:
It's used here. https://github.com/Devsh-Graphics-Programming/Nabla/pull/340/files#diff-7d65c492f5eadf5360cd59017ee15e365003cdbe97fbc29de11a40f3da117512R138 I don't like to keep this thing around too, but we need to convert back to original space.

Reply:
we can only store z axis basis vector actually.